# eBay Kleinanzeigen a section of the German eBay website

This dataset was originally scraped and uploaded to Kaggle. It is modified from the original dataset that was uploaded to Kaggle. The data is also dirtied to be more like the scraped dataset as the dataset in Kaggle is cleaned to be easier to work with.

The objective of this project is to clean the data and analyze the included used car listings. This will familiarize some of the unique benefits jupyter notebook provides for pandas

In [1]:
import numpy as np
import pandas as pd

In [2]:
autos = pd.read_csv('autos.csv', encoding = "Latin-1")

In [3]:
# See what the data looks like
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

- The column name format is hard to refer, reformatting would make it easier for later operation.
- The columns that contain date value are in timestamp format where YYYY-MM-DD HH:MM:SS

In [4]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
dateCrawled            371528 non-null object
name                   371528 non-null object
seller                 371528 non-null object
offerType              371528 non-null object
price                  371528 non-null int64
abtest                 371528 non-null object
vehicleType            333659 non-null object
yearOfRegistration     371528 non-null int64
gearbox                351319 non-null object
powerPS                371528 non-null int64
model                  351044 non-null object
kilometer              371528 non-null int64
monthOfRegistration    371528 non-null int64
fuelType               338142 non-null object
brand                  371528 non-null object
notRepairedDamage      299468 non-null object
dateCreated            371528 non-null object
nrOfPictures           371528 non-null int64
postalCode             371528 non-null int64
lastSeen              

- There are 20 columns and most of them are string
- There are some missing values


In [5]:
# Print array of the existing column names
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
# rename colume to snakecase instead of camelcase
autos.rename(columns={
    'dateCrawled' : 'date_crawled',
    'name' : 'name',
    'seller' : 'seller',
    'offerType' : 'offer_type',
    'price' : 'price',
    'abtest' : 'abtest',
    'vehicleType' : 'vehicle_type',
    'yearOfRegistration' : 'registration_year',
    'gearbox' : 'gearbox',
    'powerPS' : 'power_ps',
    'model' : 'model',
    'kilometer' : 'km',
    'monthOfRegistration' : 'registration_month',
    'fuelType' : 'fuel_type',
    'brand' : 'brand',
    'notRepairedDamage' : 'unrepaired_damange',
    'dateCreated' : 'ad_created',
    'nrOfPictures' : 'num_picture',
    'postalCode' : 'postal_code',
    'lastSeen' : 'last_seen'
}, inplace=True)

In [7]:
# check if the column names have been changed
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'km', 'registration_month', 'fuel_type', 'brand', 'unrepaired_damange',
       'ad_created', 'num_picture', 'postal_code', 'last_seen'],
      dtype='object')

The column name from Camelcase is changed to snakecase as it is Python's preferred. It is also easier when we would like to refer to the column. Using camelcase, there are more chance of making mistake as it is case sensitive and we might not get it right.

In [8]:
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_ps  model  \
0    480   test          NaN               1993    manuell         0   golf   
1  18300   test        coupe               2011    manuell       190    NaN   
2   9800   test          suv               2004  automatik       163  grand   
3   1500   test   kleinwagen               2001    manuell        75   golf   
4   3600   test   kleinwagen               2008    manuell        69  fabia   

       km  registration_month fuel_type       brand unrepaired_damange  \
0  150000                   0    benzin  volkswagen                NaN   
1  125000                   5    diesel        audi                 ja   
2  125000                   8    diesel        jeep                NaN   
3  150000                   6    benzin  volkswagen               nein   
4   90000                   7    diesel       skoda               nein   

            ad_created  num_picture  postal_code            last_seen  
0  2016-03-24 00:00:00            0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00            0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00            0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00            0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00            0        60437  2016-04-06 10:17:21

## Basic Data Exploration
Determine what other cleaning tasks need to be done.
- Look fore text columns where all or almost all values are the same. Often, it can be dropped as they do not contain useful information for analysis
- numeric data that is stored as text can be cleaned or converted
- df.describe(include='all'), series.value_counts() and series.head() - this give a closer look to the series (column)

In [9]:
# Look at some descriptive statistics for all columns
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model             km  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damange           ad_created  num_picture   postal_code  \
count              299468               371528     371528.0  371528.00000   
unique                  2                  114          NaN           NaN   
top                  nein  2016-04-03 00:00:00          NaN           NaN   
freq               263182                14450          NaN           NaN   
mean                  NaN                  NaN          0.0   50820.66764   
std                   NaN                  NaN          0.0   25799.08247   
min                   NaN                  NaN          0.0    1067.00000   
25%                   NaN                  NaN          0.0   30459.00000   
50%                   NaN                  NaN          0.0   49610.00000   
75%                   NaN                  NaN          0.0   71546.00000   
max                   NaN                  NaN          0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

#### Observations
- max registration_year = 9999? min = 1000?
- num_picture is odd, we will investigate this

In [10]:
autos["num_picture"].value_counts()

0    371528
Name: num_picture, dtype: int64

"num_picture" contains only 0 for every row, we will drop this colum

In [11]:
# drop the column
autos = autos.drop(["num_picture"],axis=1)

### Explore KM and Price

In [12]:
autos["km"].value_counts()

150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: km, dtype: int64

The value in "km" are rounded and there are more high mileage than low mileage vehicles.

Analyzing column using max and min looking for unrealistic value (outliers) with the following command:
- series.unique().shape : see how many unique values
- series.describe() to view min/max/median/mean
- series.value_counts(), with some variations:
    .head()
    This returns a series thus we can do sort by using series.sort_index() with ascending = T/F
    removing outliers df[df["col"].between(x,y)]

In [13]:
print("There are",autos["price"].unique().shape, "unique values")

There are (5597,) unique values


In [14]:
autos["price"].describe()

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

In [15]:
autos["price"].value_counts().head(20)

0       10778
500      5670
1500     5394
1000     4649
1200     4594
2500     4438
600      3819
3500     3792
800      3784
2000     3432
999      3364
750      3203
650      3150
4500     3053
850      2946
2200     2936
700      2936
1800     2886
900      2874
950      2793
Name: price, dtype: int64

- The price is also seem to be rounded
- There are 5597 unique values
- One observation to note is that there are more than 10k cars with price = $0 which is odd
- The maximum price is a bit over as well as it is 2 trillion

In [20]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

3895000     1
3890000     1
2995000     1
2795000     1
1600000     2
1300000     1
1250000     2
1234566     1
1111111     2
1010010     1
1000000     5
999999     13
999990      1
911911      1
849000      1
820000      1
780000      1
745000      2
725000      1
700000      1
Name: price, dtype: int64

In [17]:
autos["price"].value_counts().sort_index(ascending=True).head(50)

0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
16        2
17        5
18        3
19        3
20       51
21        1
24        1
25       33
26        1
27        1
29        2
30       55
32        1
33        1
35       18
38        1
39        6
40       45
45       16
47        1
49       12
50      327
55       20
58        1
59        6
60       58
65       19
66        2
69        3
70       69
74        1
75       54
77        1
79        4
80      146
Name: price, dtype: int64

- There are a number of listings with prices below \$100 and a number of listing with very high values.
- With the knowledge that eBay is an auntion site, there could be items where the opening bid is $1 thus we will keep the low values
- However, with the extremely high value above 3895000 we will consider that as unrealistic. This comes from the steady increase in price until it reach 3.8m

In [18]:
autos = autos[autos["price"].between(1,3895000)]
autos["price"].describe()

count    3.606980e+05
mean     6.068401e+03
std      1.818741e+04
min      1.000000e+00
25%      1.250000e+03
50%      3.000000e+03
75%      7.490000e+03
max      3.895000e+06
Name: price, dtype: float64

We have filtered out the rows with extreme value for price

### Exploring the date columns
There are 5 columns with the date type
- date_crawled
- registration_year : int64
- registration_month : int 64
- ad_created
- last_seen

The non registration columns are stored as string

In [24]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 360698 entries, 0 to 371527
Data columns (total 19 columns):
date_crawled          360698 non-null object
name                  360698 non-null object
seller                360698 non-null object
offer_type            360698 non-null object
price                 360698 non-null int64
abtest                360698 non-null object
vehicle_type          326592 non-null object
registration_year     360698 non-null int64
gearbox               342987 non-null object
power_ps              360698 non-null int64
model                 342414 non-null object
km                    360698 non-null int64
registration_month    360698 non-null int64
fuel_type             330772 non-null object
brand                 360698 non-null object
unrepaired_damange    293958 non-null object
ad_created            360698 non-null object
postal_code           360698 non-null int64
last_seen             360698 non-null object
dtypes: int64(6), object(13)
memory usag

In [26]:
# Show the date column with string 
autos[['date_crawled','ad_created','last_seen']][0:10]

date_crawled           ad_created            last_seen
0   2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
1   2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2   2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3   2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4   2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21
5   2016-04-04 17:36:23  2016-04-04 00:00:00  2016-04-06 19:17:07
6   2016-04-01 20:48:51  2016-04-01 00:00:00  2016-04-05 18:18:39
8   2016-04-04 23:42:13  2016-04-04 00:00:00  2016-04-04 23:42:13
9   2016-03-17 10:53:50  2016-03-17 00:00:00  2016-03-31 17:17:06
10  2016-03-26 19:54:18  2016-03-26 00:00:00  2016-04-06 10:45:34

- The first 10 characters represent the date YYYY-MM-DD 

We will extract the date values, use series.value_counts() to generate a distribution and then sort by the index

In [28]:
(autos["date_crawled"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_index()
)

2016-03-05    0.025545
2016-03-06    0.014480
2016-03-07    0.035653
2016-03-08    0.033466
2016-03-09    0.034112
2016-03-10    0.032648
2016-03-11    0.032775
2016-03-12    0.036244
2016-03-13    0.015781
2016-03-14    0.036327
2016-03-15    0.033424
2016-03-16    0.030200
2016-03-17    0.031655
2016-03-18    0.013119
2016-03-19    0.035268
2016-03-20    0.036399
2016-03-21    0.035695
2016-03-22    0.032493
2016-03-23    0.032002
2016-03-24    0.029914
2016-03-25    0.032798
2016-03-26    0.031969
2016-03-27    0.030230
2016-03-28    0.035060
2016-03-29    0.034123
2016-03-30    0.033529
2016-03-31    0.031874
2016-04-01    0.034150
2016-04-02    0.035096
2016-04-03    0.038811
2016-04-04    0.037630
2016-04-05    0.012786
2016-04-06    0.003127
2016-04-07    0.001619
Name: date_crawled, dtype: float64

In [29]:
(autos["date_crawled"]
 .str[:10]
 .value_counts(normalize=True, dropna=False)
 .sort_values()
)

2016-04-07    0.001619
2016-04-06    0.003127
2016-04-05    0.012786
2016-03-18    0.013119
2016-03-06    0.014480
2016-03-13    0.015781
2016-03-05    0.025545
2016-03-24    0.029914
2016-03-16    0.030200
2016-03-27    0.030230
2016-03-17    0.031655
2016-03-31    0.031874
2016-03-26    0.031969
2016-03-23    0.032002
2016-03-22    0.032493
2016-03-10    0.032648
2016-03-11    0.032775
2016-03-25    0.032798
2016-03-15    0.033424
2016-03-08    0.033466
2016-03-30    0.033529
2016-03-09    0.034112
2016-03-29    0.034123
2016-04-01    0.034150
2016-03-28    0.035060
2016-04-02    0.035096
2016-03-19    0.035268
2016-03-07    0.035653
2016-03-21    0.035695
2016-03-12    0.036244
2016-03-14    0.036327
2016-03-20    0.036399
2016-04-04    0.037630
2016-04-03    0.038811
Name: date_crawled, dtype: float64

It seems the site is crawled on daily basis and the distribution of listings crawlede on each day is uniform

In [30]:
(autos["last_seen"]
 .str[:10]
 .value_counts(normalize=True,dropna=False)
 .sort_index()
)

2016-03-05    0.001264
2016-03-06    0.004098
2016-03-07    0.005201
2016-03-08    0.007940
2016-03-09    0.009823
2016-03-10    0.011458
2016-03-11    0.012953
2016-03-12    0.023241
2016-03-13    0.008409
2016-03-14    0.012174
2016-03-15    0.016327
2016-03-16    0.016415
2016-03-17    0.028703
2016-03-18    0.006892
2016-03-19    0.016332
2016-03-20    0.019889
2016-03-21    0.020031
2016-03-22    0.020510
2016-03-23    0.018018
2016-03-24    0.019163
2016-03-25    0.018997
2016-03-26    0.015955
2016-03-27    0.016720
2016-03-28    0.022190
2016-03-29    0.023285
2016-03-30    0.023721
2016-03-31    0.024239
2016-04-01    0.023898
2016-04-02    0.024971
2016-04-03    0.025320
2016-04-04    0.025531
2016-04-05    0.126968
2016-04-06    0.218942
2016-04-07    0.130422
Name: last_seen, dtype: float64

With the last seen, it allows us to determine on what day the listing was removed which could be because the car was sold. The last three days of last seen value has a spike in sales

In [33]:
print(autos["ad_created"].str[:10].unique().shape)
(autos["ad_created"]
.str[:10]
.value_counts(normalize=True,dropna=False)
.sort_index()
)

(114,)


2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
2015-08-10    0.000003
2015-09-04    0.000006
2015-09-09    0.000003
2015-10-14    0.000003
2015-11-02    0.000003
2015-11-08    0.000003
2015-11-10    0.000003
2015-11-12    0.000003
2015-11-13    0.000003
2015-11-17    0.000003
2015-11-23    0.000006
2015-11-24    0.000006
2015-12-05    0.000008
2015-12-06    0.000003
2015-12-17    0.000003
2015-12-27    0.000003
2015-12-28    0.000003
2015-12-30    0.000008
2016-01-02    0.000011
2016-01-03    0.000003
2016-01-06    0.000003
2016-01-07    0.000008
2016-01-08    0.000003
2016-01-10    0.000011
2016-01-13    0.000008
                ...   
2016-03-09    0.034159
2016-03-10    0.032523
2016-03-11    0.032809
2016-03-12    0.036080
2016-03-13    0.017053
2016-03-14    0.035262
2016-03-15    0.033349
2016-03-16    0.030424
2016-03-17    0.031253
2016-03-18    0.014020
2016-03-19    0.034300
2016-03-20    0.036488
2016-03-21 

There is a large variety of ad created dates. Most fall within 1-2 months of the listing date, but a few are quite old.

In [34]:
autos["registration_year"].describe()

count    360698.000000
mean       2004.451877
std          82.110038
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

The year when the car was registered is likely to tell the age of the car. Ther are some odd values min=1000 and max=9999. The min is long before the car was invented and the max is way too far in the future.

### Dealing with incorrect registration year
a car can't be first registered before the listing was seen, thus any vehicle with a registration year above 2016 is inaccurate. We will remove the listings that contains unrealistics value by first seee what percentage of the data has invalud values in this column

In [37]:
(~autos["registration_year"].between(1900,2016)).sum()/autos.shape[0]

0.03875264071328369

There are less than 4 pecent thus we will remove these rows

In [38]:
# Remove the invalid rows
autos = autos[autos['registration_year'].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(10)

2000    0.066702
1999    0.063544
2005    0.062670
2006    0.057704
2001    0.056948
2003    0.056547
2004    0.056169
2002    0.054280
2007    0.050490
1998    0.049691
Name: registration_year, dtype: float64

Most vehicles were first registered in the past 20 years

### Exploring price by Brand

In [42]:
autos["brand"].value_counts(normalize=True).head(20)

volkswagen        0.211675
bmw               0.109870
opel              0.106397
mercedes_benz     0.096833
audi              0.089536
ford              0.068912
renault           0.047508
peugeot           0.030148
fiat              0.025686
seat              0.018658
skoda             0.015684
mazda             0.015381
smart             0.014329
citroen           0.013948
nissan            0.013596
toyota            0.012930
hyundai           0.009971
sonstige_autos    0.009547
mini              0.009382
volvo             0.009146
Name: brand, dtype: float64

The German manufacturers represent the 4/5 top brands and almost 50% of the total listings. Volkswagen is the most popular brand with 21%

There are a lot of brand that dont have a significant percantage listings, we will limit our analysis to those with more than 5% of total listings.

In [43]:
brand_counts = autos["brand"].value_counts(normalize=True)
common_brands = brand_counts[brand_counts > 0.05].index
print(common_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [49]:
brand_mean_price ={}

for brand in common_brands:
    brand_only = autos[autos["brand"]==brand]
    mean_price = brand_only["price"].mean()
    brand_mean_price[brand] = int(mean_price)

brand_mean_price  

{'volkswagen': 5452,
 'bmw': 8573,
 'opel': 3028,
 'mercedes_benz': 8609,
 'audi': 9166,
 'ford': 3779}

There are price gap between the top 5 brands
- audi, bmw, and mercedes are the expensice ones
- Opel and ford are less expensive
- volkswagen is in between, this could be the reason why it is most popular

### Exploring the Mileage

In [51]:
# Putting the dictionary we previously created into a dataframe
bmp_series = pd.Series(brand_mean_price)
pd.DataFrame(bmp_series,columns=["mean_mileage"])

mean_mileage
volkswagen             5452
bmw                    8573
opel                   3028
mercedes_benz          8609
audi                   9166
ford                   3779

In [52]:
brand_mean_mileage = {}

for brand in common_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["km"].mean()
    brand_mean_mileage[brand]= int(mean_mileage)
    
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_price=pd.Series(brand_mean_price).sort_values(ascending=False)

In [53]:
brand_info = pd.DataFrame(mean_mileage,columns=["mean_mileage"])
brand_info

mean_mileage
bmw                  132791
mercedes_benz        130566
audi                 129440
opel                 128721
volkswagen           128387
ford                 123663

In [54]:
brand_info["mean_price"] = mean_price
brand_info

mean_mileage  mean_price
bmw                  132791        8573
mercedes_benz        130566        8609
audi                 129440        9166
opel                 128721        3028
volkswagen           128387        5452
ford                 123663        3779

The range of car mileages does not vary as much as the price across the brands. The expensive ones tends to have higher mileage while the less expensive are having lower mileage